packages/network/src/network.js

```diff
@@ -27,554 +27,759 @@ export const rethrowUnlessMissing = err => {
+) => {
+  const detached = zone.detached();
+
+  /** @type {MapStore<number, Connection>} */
+  const conns = detached.mapStore('addrToConnections');
+
+  /** @type {ConnectionHandler[]} */
+  const handlers = harden([handler0, handler1]);
+  /** @type {Endpoint[]} */
+  const addrs = harden([addr0, addr1]);
+
+  const makeHalfConnection = (l, r) => {
+    conns.init(l, makeConnection({ addrs, handlers, conns, current, l, r }));
+  };
 
   /**
    * @param {number} l local side of the connection
    * @param {number} r remote side of the connection
    */
-  function openHalfConnection(l, r) {
-    current.add(conns[l]);
+  const openHalfConnection = (l, r) => {
+    current.add(conns.get(l));
     E(handlers[l])
-      .onOpen(conns[l], addrs[l], addrs[r], handlers[l])
+      .onOpen(conns.get(l), addrs[l], addrs[r], handlers[l])
       .catch(rethrowUnlessMissing);
-  }
+  };
+
+  makeHalfConnection(0, 1);
+  makeHalfConnection(1, 0);
 
   openHalfConnection(0, 1);
   openHalfConnection(1, 0);
 
-  const [conn0, conn1] = conns;
-  return [conn0, conn1];
-}
+  return [conns.get(0), conns.get(1)];
+};
 
-/**
- * Get the list of prefixes from longest to shortest.
- *
- * @param {string} addr
- */
-export function getPrefixes(addr) {
-  const parts = addr.split(ENDPOINT_SEPARATOR);
+const prepareInboundAttempt = (zone, makeConnection) => {
+  const makeInboundAttempt = zone.exoClass(
+    'InboundAttempt',
+    undefined,
+    /**
+     * @param {object} opts
+     * @param {string} opts.localAddr
+     * @param {string} opts.remoteAddr
+     * @param { MapStore<Port, SetStore<Closable>> } opts.currentConnections
+     * @param {string} opts.listenPrefix
+     * @param {MapStore<Endpoint, [Port, ListenHandler]>} opts.listening
+     */
+    ({
+      localAddr,
+      remoteAddr,
+      currentConnections,
+      listenPrefix,
+      listening,
+    }) => {
+      /** @type {Error | undefined} */
+      let consummated;
 
-  /** @type {string[]} */
-  const ret = [];
-  for (let i = parts.length; i > 0; i -= 1) {
-    // Try most specific match.
-    const prefix = parts.slice(0, i).join(ENDPOINT_SEPARATOR);
-    ret.push(prefix);
-  }
-  return ret;
-}
+      return {
+        localAddr,
+        remoteAddr,
+        consummated,
+        currentConnections,
+        listenPrefix,
+        listening,
+      };
+    },
+    {
+      getLocalAddress() {
+        // Return address metadata.
+        return this.state.localAddr;
+      },
+      getRemoteAddress() {
+        return this.state.remoteAddr;
+      },
+      async close() {
+        if (this.state.consummated) {
+          throw this.state.consummated;
+        }
+        this.state.consummated = Error(`Already closed`);
 
-/**
- * Create a protocol that has a handler.
- *
- * @param {ProtocolHandler} protocolHandler
- * @returns {Protocol} the local capability for connecting and listening
- */
-export function makeNetworkProtocol(protocolHandler) {
```

Comment:
I'm looking at some dapp code that uses `makeNetworkProtocol` and I realize: deleting this exported function is a **breaking API change**.

Reply:
Hmm... I wonder what the right way is to go about this change. Ideally, they'd go to using the prepare version

Reply:
I asked around and learned that breaking API changes are OK/expected in this work.

Just make sure to discuss them thoroughly in the commit message(s) that go on master, including BREAKING CHANGE and/or ! as in [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)